Web Audio glitches when hardware buffer size is not divisible by 128

mwyrzykowski opened the pull request:
#### f35686d638221a75d5450d3251031dce9a9a8192
<pre>
Web Audio glitches when hardware buffer size is not divisible by 128
<a href="https://bugs.webkit.org/show_bug.cgi?id=259899">https://bugs.webkit.org/show_bug.cgi?id=259899</a>
&lt;radar://112621241&gt;

Reviewed by Chris Dumez.

Whenever the buffer size of the hardware was not a multiple of 128, the
audio would not be in sync with the client leading to audible artifacts.

We can fix this by passing the number of frames from AURemoteIO to the
web process and rendering all of the frames that AURemoteIO asked us to render.

* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
(WebKit::RemoteAudioDestinationManager::createAudioDestination):
Create a shared memory member variable which we can write the frame count from
AURemoteUI to.

* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.h:
Update member function parameters.

* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.messages.in:
Pass the shared memory handle to the web process.

* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::startRenderingThread):
Get the number of frames to render from the shared memory with the GPU process.

(WebKit::RemoteAudioDestinationProxy::connection):
Create the shared memory handle in the web process and pass it to the GPU process.

We could create it the other way around, but it would require adding another IPC
message, which seems uncessary for this purpose.

* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
Add shared memory member variable.

Canonical link: <a href="https://commits.webkit.org/266747@main">https://commits.webkit.org/266747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ccbbe4984f64b13872aa39b8fbfff7158397de68

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/14715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14949 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15297 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16387 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/13824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17466 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15032 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16470 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14820 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15331 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17121 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12614 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20208 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13693 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13370 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16607 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13923 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13217 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3529 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17555 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13766 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
```diff
@@ -142,12 +143,9 @@ class RemoteAudioDestination final
-        }
-        m_extraRequestedFrames = requestedSamplesCount - numberOfFrames;
+        ASSERT(m_frameCount->size() == sizeof(numberOfFrames));
+        memcpy(m_frameCount->data(), &numberOfFrames, sizeof(numberOfFrames));
```
@kkinnunen-apple should probably comment but this doesn't look safe to me because it doesn't look like anything prevents the GPUProcess from writing into the frameCount shared memory while the WebProcess is in the middle of reading it. As such, the WebProcess could read corrupted data in some cases.

I believe we use `Atomics` to avoid such issues in other places in our code base. E.g.
```
struct Header {
        Atomic<ServerOffset> serverOffset;
        // Padding so that the variables mostly accessed by different processes do not share a cache line.
        // This is an attempt to avoid cache-line induced reduction of parallel access.
        // Use 128 bytes since that's the cache line size on ARM64, and enough to cover other platforms where 64 bytes is common.
        alignas(128) Atomic<ClientOffset> clientOffset;
    };

Header& header() const { return *reinterpret_cast<Header*>(m_sharedMemory->data()); }
```

That's from our streaming IPC implementation that Kimmo wrote so he is likely a good person to ask for help here.


Another more high level concern which may or may not be an issue in practice:
1. GPUProcess writes 128
2. GPUProcess writes 256
3. GPUProcess writes 128

but the WebProcess is a bit slow and took too long to process 1., as a result, by the time it processes the second quantum, 256 has already been overwritten by 128. So the WebProcess reads 128, 128, 128 and doesn't produce enough samples.

mwyrzykowski replied:
Regarding corruption, I am relying on the fact that aligned memory won't be written from more than one thread and not at less than sizeof(register) at a time, so there should be no corrupted reads. But indeed the higher level race certainly exists.

mwyrzykowski replied:
Atomic<> seems more appropriate in any case, thank you for alerting me of that, I was unaware of it

mwyrzykowski replied:
We could add an IPC message from RemoteAudioDestination to RemoteAudioDestinationProxy containing the number of frames to read to address the higher level race but I'm not sure if that would add latency to the audio processing.

cdumez replied:
Not sure if that would work but if you use `Atomic`, you might be able to:
1. Have the GPUProcess increment frameCount for every render call (instead of overwriting the existing value)
2. Have the WebProcess exchange the frameCount with 0 atomically and use that frameCount to decide how much to render.

mwyrzykowski replied:
That's a great idea. I'll try that.

mwyrzykowski replied:
@cdumez I effectively took this suggestion, but instead of exchanging with 0 in the web process (which would involve writing to the shared memory from both processes), I kept a counter and handle overflow which should work out to be equivalent. This way only the GPU Process writes to the shared memory as I was not sure using atomic exchange on a shared memory location is safe.

I think this addresses the race, please let me know what you think.

cdumez replied:
Yes, this way of dealing with the race seems OK to me.

It'd be good to switch to Atomic though.

mwyrzykowski replied:
Great, I updated the PR to directly use `WTF::atomicLoad` and `WTF::atomicExchangeAdd` which calls into `WTF::Atomic` and also added a static assert that the atomic operations on uint32_t are lock free otherwise as far as I know they are not guaranteed to be atomic for shared memory

cdumez commented on Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
```diff
@@ -79,11 +79,9 @@ void RemoteAudioDestinationProxy::startRenderingThread()
-            while (m_renderSemaphore.waitFor(0_s))
-                frameCount += WebCore::AudioUtilities::renderQuantumSize;
+            ASSERT(m_frameCount->size() == sizeof(uint32_t));
+            renderAudio(*static_cast<uint32_t*>(m_frameCount->data()));
```
Not that even though you may now ask for a number of frames that is not a multiple of 128, our lower level code will STILL produce a number of samples that is a multiple of 128 (thus producing too many samples):
```
void AudioDestinationResampler::renderOnRenderingThreadIfPlaying(size_t framesToRender)
{
    if (!m_isPlaying)
        return;
    for (size_t pushedFrames = 0; pushedFrames < framesToRender; pushedFrames += AudioUtilities::renderQuantumSize) {
        if (m_resampler)
            m_resampler->process(m_renderBus.ptr(), AudioUtilities::renderQuantumSize);
        else
            callRenderCallback(nullptr, m_renderBus.ptr(), AudioUtilities::renderQuantumSize, m_outputTimestamp);
```

`AudioUtilities::renderQuantumSize` is 128.

cdumez replied:
Maybe that's fine but it makes it less obvious what actually changed with this patch.

eric-carlson commented on Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
```diff
@@ -142,12 +143,9 @@ class RemoteAudioDestination final
-            m_renderSemaphore.signal();
-        }
-        m_extraRequestedFrames = requestedSamplesCount - numberOfFrames;
+        ASSERT(m_frameCount->size() == sizeof(numberOfFrames));
```
Should this be a release assert?

eric-carlson commented on Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
```diff
@@ -79,11 +79,12 @@ void RemoteAudioDestinationProxy::startRenderingThread()
-            while (m_renderSemaphore.waitFor(0_s))
-                frameCount += WebCore::AudioUtilities::renderQuantumSize;
-
+            ASSERT(m_frameCount->size() == sizeof(uint32_t));
```
Ditto

cdumez commented on Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
```diff
@@ -142,12 +143,10 @@ class RemoteAudioDestination final
-        m_extraRequestedFrames = requestedSamplesCount - numberOfFrames;
+        static_assert(std::atomic<UInt32>::is_always_lock_free, "Shared memory atomic usage assumes lock free primitives are used");
+        RELEASE_ASSERT(m_frameCount->size() == sizeof(numberOfFrames));
+        WTF::atomicExchangeAdd(static_cast<uint32_t*>(m_frameCount->data()), numberOfFrames);
```
This ends up casting the SharedMemory pointer into a `WTF::Atomics<uint32_t>*`, which wraps a `std::atomics<uint32_t>`. Your SharedMemory has `size(uint32_t)`. What guarantees that this is the same size as `std::atomics<uint32_t>`? As far as I know there is no such guarantee so this usually is the case.

mwyrzykowski replied:
This is a good point, I updated the PR to allocate sizeof(atomic<uint32_t>) shared memory.

We should probably also update `atomicExchangeAdd` and similar functions with static asserts, that is outside the scope of this PR, I can file an issue tracking it

cdumez commented on Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
```diff
@@ -142,12 +143,10 @@ class RemoteAudioDestination final
-            m_renderSemaphore.signal();
-        }
-        m_extraRequestedFrames = requestedSamplesCount - numberOfFrames;
+        static_assert(std::atomic<UInt32>::is_always_lock_free && sizeof(std::atomic<uint32_t>) == sizeof(UInt32), "Shared memory atomic usage assumes lock free primitives are used");
```
I don’t think we need the second part of this assertion (the size check).

cdumez reviewed (approved):
LGTM with nit.

cdumez reviewed (approved):
Still LGTM after changes.

mwyrzykowski committed:
Web Audio glitches when hardware buffer size is not divisible by 128
https://bugs.webkit.org/show_bug.cgi?id=259899
<radar://112621241>

Reviewed by Chris Dumez.

Whenever the buffer size of the hardware was not a multiple of 128, the
audio would not be in sync with the client leading to audible artifacts.

We can fix this by passing the number of frames from AURemoteIO to the
web process and rendering all of the frames that AURemoteIO asked us to render.

* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
(WebKit::RemoteAudioDestinationManager::createAudioDestination):
Create a shared memory member variable which we can write the frame count from
AURemoteUI to.

* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.h:
Update member function parameters.

* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.messages.in:
Pass the shared memory handle to the web process.

* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::startRenderingThread):
Get the number of frames to render from the shared memory with the GPU process.

(WebKit::RemoteAudioDestinationProxy::connection):
Create the shared memory handle in the web process and pass it to the GPU process.

We could create it the other way around, but it would require adding another IPC
message, which seems uncessary for this purpose.

* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
Add shared memory member variable.

Canonical link: https://commits.webkit.org/266747@main
